File: backend/services/ai-service/src/modules/ai/gemini.client.ts
import { GoogleGenerativeAI } from '@google/generative-ai';
import config from '../../config';

let _client: GoogleGenerativeAI | null = null;

function getClient(): GoogleGenerativeAI | null {
  if (!config.gemini.apiKey) return null;
  if (!_client) _client = new GoogleGenerativeAI(config.gemini.apiKey);
  return _client;
}

export async function geminiGenerate(prompt: string): Promise<string | null> {
  const client = getClient();
  if (!client) return null;

  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), 10_000);

  try {
    const model = client.getGenerativeModel({ model: config.gemini.model });
    const result = await model.generateContent({
      contents: [{ role: 'user', parts: [{ text: prompt }] }],
      generationConfig: {
        responseMimeType: 'application/json',
        temperature: 0.3,
        maxOutputTokens: 2048,
      },
    });
    clearTimeout(timer);
    return result.response.text() ?? null;
  } catch {
    clearTimeout(timer);
    return null;
  }
}

/**
 * Stream tokens from Gemini, calling onChunk for each text piece.
 * Uses plain text output (not JSON) — suitable for SSE streaming.
 */
export async function geminiStream(
  prompt: string,
  onChunk: (text: string) => void,
  signal?: AbortSignal,
): Promise<void> {
  const client = getClient();
  if (!client) throw new Error('Gemini not configured');

  const abortController = new AbortController();
  if (signal) signal.addEventListener('abort', () => abortController.abort());
  const timer = setTimeout(() => abortController.abort(), 90_000);

  try {
    const model = client.getGenerativeModel({ model: config.gemini.model });
    const result = await model.generateContentStream({
      contents: [{ role: 'user', parts: [{ text: prompt }] }],
      generationConfig: { temperature: 0.4, maxOutputTokens: 2048 },
    });

    for await (const chunk of result.stream) {
      if (abortController.signal.aborted) break;
      const text = chunk.text();
      if (text) onChunk(text);
    }
  } finally {
    clearTimeout(timer);
  }
}

export function isGeminiConfigured(): boolean {
  return Boolean(config.gemini.apiKey);
}

File: backend/services/ai-service/app/clients/gemini_client.py
from collections.abc import AsyncIterator

import google.generativeai as genai

from app.config import settings

_model = None


def _get_model():
    global _model
    if not settings.gemini_api_key:
        return None
    if _model is None:
        genai.configure(api_key=settings.gemini_api_key)
        _model = genai.GenerativeModel(settings.gemini_model)
    return _model


async def gemini_generate(prompt: str) -> str | None:
    model = _get_model()
    if model is None:
        return None
    try:
        response = model.generate_content(
            prompt,
            generation_config=genai.types.GenerationConfig(
                response_mime_type="application/json",
                temperature=0.3,
                max_output_tokens=2048,
            ),
        )
        return response.text
    except Exception:
        return None


async def gemini_stream(prompt: str) -> AsyncIterator[str]:
    """Yield tokens from Gemini streaming (plain text, not JSON)."""
    model = _get_model()
    if model is None:
        raise RuntimeError("Gemini not configured")

    response = model.generate_content(
        prompt,
        stream=True,
        generation_config=genai.types.GenerationConfig(
            temperature=0.4,
            max_output_tokens=2048,
        ),
    )
    for chunk in response:
        text = chunk.text
        if text:
            yield text


def is_gemini_configured() -> bool:
    return bool(settings.gemini_api_key)
